### ISSUE TITLE
Ray Executor

### ISSUE THREAD

**Issue by alxmrs** [cubed-dev/cubed#488]
In addition to accelerator support (e.g. via #304), Cubed could benefit ML users by providing [ray](https://github.com/ray-project/ray) executor: https://docs.ray.io/en/latest/ray-core/walkthrough.html

Since Cubed is a serverless model, I bet it could get away with only using [Tasks/remote functions](https://docs.ray.io/en/latest/ray-core/tasks.html#ray-remote-functions). 

From talking with @cromwellian a bit, my hope is that Cubed could provide memory bounds when trying to saturate GPUs during model training. I'm not totally sure exactly what a training loop with Cubed would look like. Here's how ray integrates with PyTorch, for example:  https://docs.ray.io/en/latest/train/api/doc/ray.train.torch.TorchTrainer.html#ray.train.torch.TorchTrainer

@shoyer pointed out to me once the idea that GPU OOM errors occur while taking the gradient of a function graph, not necessarily on the forward pass. I'm not totally sure right now if Cubed is in fact a good fit for tackling this problem, only that the potential is exciting.

**Comment by tomwhite** [comment]
Thanks for opening this issue @alxmrs! I think Ray would be a great runtime for Cubed, and should be relatively straightforward to write an executor for (maybe a bit like the [Modal one](https://github.com/cubed-dev/cubed/blob/7f3f390925ce0b290629aed76941a84a14666d8d/cubed/runtime/executors/modal.py)?). Do you know what people generally run Ray on in production/at scale?

**Comment by alxmrs** [comment]
Hey Tom! Do you mean what does the userbase look like, or do I know specific people? On the former: Ray is the engine that OpenAI uses to train its GPT models; [it's really popular in the ML world](https://www.meetup.com/Bay-Area-Ray-Meetup/). On the latter: Ray, the person (cromwellian), uses Ray, the framework, at Roblox for model training. :)

> should be relatively straightforward to write an executor for (maybe a bit like the [Modal one](https://github.com/cubed-dev/cubed/blob/7f3f390925ce0b290629aed76941a84a14666d8d/cubed/runtime/executors/modal.py)?).

I agree, and it does look like it will be similar to Modal.

**Comment by tomwhite** [comment]
I meant usage of Anyscale vs KubeRay vs ?? I was wondering if there was a choice that most people use, or whether it's a bit of everything.

> Ray, the person (cromwellian), uses Ray, the framework, at Roblox for model training. :)

Got it!

**Comment by tomwhite** [comment]
I added some notes on how to write a new executor in #498.

**Comment by rbavery** [comment]
> I meant usage of Anyscale vs KubeRay vs ?? I was wondering if there was a choice that most people use, or whether it's a bit of everything.

From attending Ray Summit last year, I think both Anyscale and KubeRay Operator (open source, self managed Ray) are popular but if I had to guess based on conversations and talks I attended, I think there are more users of self managed Kuberay.

**Comment by tomwhite** [comment]
I've now run on a small Ray cluster on AWS: see #769.

**Comment by alxmrs** [comment]
Great work closing this, Tom!
